### ISSUE TITLE
IN predicate supports only constant list of values

### ISSUE THREAD

**Issue by karteekmurthys** [facebookincubator/velox#5894]
### Description

The following query has a `FieldTypedExpression` in the values list of IN expression:
```
presto:kart> SELECT (SELECT 2) IN (1, (SELECT 2))
          -> ;
Query 20230727_222608_00000_mbxye failed:  IN predicate supports only constant list of values
```
Velox expects the arguments to be all constants.

Relevant code:
```
for (auto i = 0; i < size; ++i) {
    auto constant =
        dynamic_cast<const ConstantTypedExpr*>((*(begin + i)).get());
    if (constant == nullptr) {
      VELOX_UNSUPPORTED("IN predicate supports only constant list of values");
    }
```

Stacktrace:
```   __cxa_throw 0x000000019667ac54
facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxUserError, const char *>(const facebook::velox::detail::VeloxCheckFailArgs &, const char *) Exceptions.h:76
facebook::presto::toArrayVector<(facebook::velox::TypeKind)3>(std::__wrap_iter<const std::shared_ptr<const facebook::velox::core::ITypedExpr> *>, std::__wrap_iter<const std::shared_ptr<const facebook::velox::core::ITypedExpr> *>, facebook::velox::memory::MemoryPool *) PrestoToVeloxExpr.cpp:548
$_0::operator()() const PrestoToVeloxExpr.cpp:584
facebook::presto::convertInExpr(const std::vector<std::shared_ptr<const facebook::velox::core::ITypedExpr>, std::allocator<std::shared_ptr<const facebook::velox::core::ITypedExpr> > > &, facebook::velox::memory::MemoryPool *) PrestoToVeloxExpr.cpp:584
facebook::presto::VeloxExprConverter::toVeloxExpr(std::shared_ptr<facebook::presto::protocol::SpecialFormExpression>) const PrestoToVeloxExpr.cpp:642
facebook::presto::VeloxExprConverter::toVeloxExpr(std::shared_ptr<facebook::presto::protocol::RowExpression>) const PrestoToVeloxExpr.cpp:705
facebook::presto::getProjections(const facebook::presto::VeloxExprConverter &, const facebook::presto::protocol::Assignments &) PrestoToVeloxQueryPlan.cpp:823
facebook::presto::VeloxQueryPlanConverterBase::toVeloxQueryPlan(const std::shared_ptr<const facebook::presto::protocol::ProjectNode> &, const std::shared_ptr<facebook::presto::protocol::TableWriteInfo> &, const std::basic_string<char, std::char_traits<char>, std::allocator<char> > &) PrestoToVeloxQueryPlan.cpp:1598
facebook::presto::VeloxQueryPlanConverterBase::toVeloxQueryPlan(const std::shared_ptr<const facebook::presto::protocol::PlanNode> &, const std::shared_ptr<facebook::presto::protocol::TableWriteInfo> &, const std::basic_string<char, std::char_traits<char>, std::allocator<char> > &) PrestoToVeloxQueryPlan.cpp:2401
facebook::presto::VeloxQueryPlanConverterBase::toVeloxQueryPlan(const std::shared_ptr<const facebook::presto::protocol::OutputNode> &, const std::shared_ptr<facebook::presto::protocol::TableWriteInfo> &, const std::basic_string<char, std::char_traits<char>, std::allocator<char> > &) PrestoToVeloxQueryPlan.cpp:2683
facebook::presto::VeloxQueryPlanConverterBase::toVeloxQueryPlan(const facebook::presto::protocol::PlanFragment &, const std::shared_ptr<facebook::presto::protocol::TableWriteInfo> &, const std::basic_string<char, std::char_traits<char>, std::allocator<char> > &) PrestoToVeloxQueryPlan.cpp:2532
proxygen::RequestHandlerAdaptor::onEOM() 0x00000001053f9a74
proxygen::HTTPTransaction::processIngressEOM() 0x00000001055677f8
proxygen::HTTPTransaction::onIngressEOM() 0x00000001055674ec
proxygen::HTTPSession::onMessageComplete(unsigned long long, bool) 0x000000010553affc
proxygen::PassThroughHTTPCodecFilter::onMessageComplete(unsigned long long, bool) 0x00000001054ccfbc
proxygen::HTTP1xCodec::onMessageComplete() 0x000000010549b15c
proxygen::HTTP1xCodec::onMessageCompleteCB(proxygen::http_parser*) 0x0000000105492e40
proxygen::http_parser_execute_options(proxygen::http_parser*, proxygen::http_parser_settings const*, unsigned char, char const*, unsigned long) 0x0000000105468810
proxygen::HTTP1xCodec::onIngressImpl(folly::IOBuf const&) 0x0000000105493ad0
proxygen::HTTP1xCodec::onIngress(folly::IOBuf const&) 0x00000001054936a8
```

On java side, the values list must all be SCALAR at least if not constants:
```
Query 20230727_234750_00004_cqjz9, FAILED, 1 node
Splits: 34 total, 1 done (2.94%)
[Latency: client-side: 0:02, server-side: 0:01] [5 rows, 787B] [3 rows/s, 564B/s]

Query 20230727_234750_00004_cqjz9 failed: Scalar sub-query has returned multiple rows

presto:kart> SELECT (SELECT 9) IN (1, (SELECT 4+5));
 _col0 
-------
 true  
(1 row)

presto:kart> SELECT (SELECT 9) IN (1, (SELECT decimal'9.0'));
 _col0 
-------
 true  
(1 row)

```

**Comment by karteekmurthys** [comment]
cc: @aditi-pandit

**Comment by aditi-pandit** [comment]
@karteekmurthys : Can you print the EXPLAIN plan here for these. Also attach the plan fragment from Presto Console for it.

**Comment by karteekmurthys** [comment]
```
 Fragment 0 [SINGLE]                                                                                                                    
     Output layout: [expr_3]                                                                                                            
     Output partitioning: SINGLE []                                                                                                     
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                      
     - Output[_col0] => [expr_3:boolean]                                                                                                
             Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 12.00, memory: 0.00, network: 0.00}                            
             _col0 := expr_3 (1:47)                                                                                                     
         - Project[projectLocality = LOCAL] => [expr_3:boolean]                                                                         
                 Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 12.00, memory: 0.00, network: 0.00}                        
                 expr_3 := IN(expr, INTEGER'1', expr) (1:47)                                                                            
             - LocalExchange[ROUND_ROBIN] () => [expr:integer]                                                                          
                     Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 10.00, memory: 0.00, network: 0.00}                    
                 - EnforceSingleRow => [expr:integer]                                                                                   
                         Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 5.00, memory: 0.00, network: 0.00}                 
                     - LocalExchange[SINGLE] () => [expr:integer]                                                                       
                             Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 5.00, memory: 0.00, network: 0.00}             
                         - Project[projectLocality = LOCAL] => [expr:integer]                                                           
                                 Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 5.00, memory: 0.00, network: 0.00}         
                                 expr := INTEGER'2'                                                                                     
                             - LocalExchange[ROUND_ROBIN] () => []                                                                      
                                     Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 0.00, memory: 0.00, network: 0.00}     
                                 - Values => []                                                                                         
                                         Estimates: {source: CostBasedSourceInfo, rows: 1 (2B), cpu: 0.00, memory: 0.00, network: 0.00} 
                                         () 
```

**Comment by karteekmurthys** [comment]
```
"assignments" : {
            "assignments" : {
              "expr_3\u003Cboolean\u003E" : {
                "@type" : "special",
                "sourceLocation" : {
                  "line" : 1,
                  "column" : 19
                },
                "form" : "IN",
                "returnType" : "boolean",
                "arguments" : [ {
                  "@type" : "variable", <----------- LHS to IN is a VariableExpression
                  "sourceLocation" : {
                    "line" : 1,
                    "column" : 19
                  },
                  "name" : "expr",
                  "type" : "integer"
                }, {
                  "@type" : "constant",
                  "valueBlock" : "CQAAAElOVF9BUlJBWQEAAAAAAQAAAA==",
                  "type" : "integer"
                }, {
                  "@type" : "variable", <--- The IN list has variable expression.
                  "sourceLocation" : {
                    "line" : 1,
                    "column" : 19
                  },
                  "name" : "expr",
                  "type" : "integer"
                } ]
              }
```

**Comment by mknegi** [comment]
Another e2d testcase that counters this problem:
**Testcase:** presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java::testIn()
**Failing Query:** SELECT x FROM (values 1, 2, 3) t(x) WHERE x IN (1 + CAST(rand() < 0 AS bigint), 2 + CAST(rand() < 0 AS bigint))
**Error reported:**
VeloxUserError:  IN predicate supports only constant list of values

**Comment by mbasmanova** [comment]
Fixed in #7497 and #7585
